[HEAP-28592] improve error in looker

yiyangli-heap opened the pull request:
- include the form validation in async form, the method is called when the configuration is saved;
- don't throw an error in onFields and onRow, instead, put error to errors. if we want to skip a row, don't push error to errors;
- if errors is not empty, return the failure response
- for every info, error log, try to include envId and webhookId. The webhookId can help us categorize logs for a job.

HEAP-28592

yiyangli-heap committed:
[HEAP-28592] improve error improvement

- include the form validation in async form, the method is called when the configuration is saved;
- don't throw an error in onFields and onRow, instead, put error to errors. if we want to skip a row, don't push error to errors;
- if errors is not empty, return the failure response
- for every info, error log, try to include envId and webhookId. The webhookId can help us categorize logs for a job.

HEAP-28592

yiyangli-heap committed:
fix lint and add tests

yiyangli-heap commented on src/actions/heap/heap.ts:
```diff
@@ -136,25 +150,27 @@ export class HeapAction extends Hub.Action {
           }
         } catch (err) {
-          errors.push(err)
+          errors.push(new Error("Encountered an error onRow, error: " + getErrorMessage(err)))
```
I want to create an InternalError here, and filter them out whent he final response is generated.

yiyangli-heap commented on src/actions/heap/heap.ts:
```diff
@@ -165,29 +181,46 @@ export class HeapAction extends Hub.Action {
+      // log first N errors
+      errorsToDisplay.forEach((err) => winston.error(`envId ${envId} error: ${err.message}`, logTag))
+      // concat first N errors into a signle errorMsg to return to the looker action hub.
+      const errorMsg = errorsToDisplay.map((err) => err.message).join(", ")
```
filter out InternalError in the next PR. I want to deploy this version and see the response firstly, see if it's improved.

yiyangli-heap commented on src/actions/heap/heap.ts:
```diff
@@ -267,77 +307,112 @@ export class HeapAction extends Hub.Action {
+     const value = row[fieldName].value != null ? row[fieldName].value.toString() : ""
+     return {
         key: fieldName,
-        value: row[fieldName].value,
```
I think this is why the error is 500, fieldName exists, but value is undefined. and next, value.toString is called.

colbysadams replied:
value.toString is only called after we check if the returned value was falsy... rather than attaching the error to your returned value here, I think we should throw an error when the value is undefined and then handle that error in the calling function.

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -58,19 +70,13 @@ export class HeapAction extends Hub.Action {
   supportedActionTypes = [Hub.ActionType.Query]
   supportedPropertyTypes = [HeapPropertyTypes.User, HeapPropertyTypes.Account]
   usesStreaming = true
-  supportedFormats = (request: Hub.ActionRequest) => {
```
why was this removed?

yiyangli-heap replied:
I don't think it's used. I checked other actions. 

https://github.com/looker/actions/blob/e2d0d8c0261bc06d48b2a98e4327a80dcbc5ef50/src/actions/google/drive/google_drive.ts#L16-L21

https://github.com/heap/actions/blob/master/src/actions/slack/slack.ts#L23-L33

Should I add a `minimumSupportedLookerVersion`? Frankly, I don't find a dev doc.

colbysadams replied:
https://docs.looker.com/sharing-and-publishing/action-hub#supported_data_formats

yiyangli-heap replied:
sounds good, I put it back.

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -136,25 +150,27 @@ export class HeapAction extends Hub.Action {
       },
     })
 
-    if (requestBatch.length > 0) {
```
why was this removed?

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -136,25 +150,27 @@ export class HeapAction extends Hub.Action {
+      errors.push(new Error("Encountered an error in execute, error: " + getErrorMessage(err)))
+    }
+
+    if (requestBatch.length > 0) {
+      const length = requestBatch.length
+      winston.info(`Loading the remaining ${length} rows of data to heap`, logTag)
+      await this.sendRequest(requestBatch, envId, requestUrl, heapField, errors)
     }
```
if we call this after the above try statement, I don't think that this error won't be caught in the `await Promise.all(requestPromises)`

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -50,6 +61,7 @@ export class HeapAction extends Hub.Action {
   static LOG_PROGRESS_STEP = 10000
   static DISPLAY_ERROR_COUNT = 250
 
+  emptyHeapEntity = { heapFieldValue: undefined, properties: {} }
```
can we make this static?

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -216,38 +249,45 @@ export class HeapAction extends Hub.Action {
   }
 
-  private extractHeapFieldByLabel(fields: Hub.Field[], heapFieldLabel: string): Hub.Field {
+  private extractHeapFieldByLabel(fields: Hub.Field[], heapFieldLabel: string, errors: Error[]): Hub.Field | undefined {
```
instead of passing in an errors array, could we instead update the calling code to catch this error when it is thrown and take the appropriate action there?

yiyangli-heap replied:
it's part of `onFields`, a callback when the stream is processed. I think what I could do is that, skip onRow when ~`heapFieldLabel`~ `identityField` is empty.

yiyangli-heap replied:
the `errors` will be bubbled up to the final part that generates the response.

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -386,8 +461,8 @@ export class HeapAction extends Hub.Action {
     } catch (err) {
-      errors.push(err)
-    }
+      errors.push(new Error("Encountered an error in sending request to heap, error: " + getErrorMessage(err)))
```
Can we make this a debug log?

yiyangli-heap replied:
It will be logged in the end, when the whole thing is completed. The original test verifies the error. I want to make it an InternalError, so that the error is not included in the final response.

yiyangli-heap committed:
call heap track even when identityField is undefined

change 2 properties to static

yiyangli-heap committed:
catch errors in onFields

colbysadams commented on src/actions/heap/heap.ts:
```diff
@@ -82,47 +94,62 @@ export class HeapAction extends Hub.Action {
+        }
       },
       onRow: (row) => {
+        if (!identityField) {
```
why are we silently dropping rows if the identity field is not set?

yiyangli-heap replied:
`identityField` is a required field for each row. If it's empty, heapFieldValue is empty, it will quit either way. choosing to quit early because `extractRequiredField` will throw an error if the value is empty.

yiyangli-heap committed:
- rename errors to displayErrors to clarify
- log error when it's added to displayErrors
- continue to count rowsReceived even if identityField is empty, extractPropertiesFromRow will swallow the error if heapFieldValue is empty

HEAP-28592
